### PR TITLE
Input definition

### PR DESCRIPTION
**Pull request by linhvo**
## Overview

This PR cover #607

## Pull request checklist

- [x] I have read the [contributing guide](https://github.com/pilosa/pilosa/blob/master/CONTRIBUTING.md).
- [x] I have agreed to the [Contributor License Agreement](https://cla-assistant.io/pilosa/pilosa).
- [x] I have updated the [documentation](https://github.com/pilosa/pilosa/tree/master/docs).
- [x] I have resolved any merge conflicts.
- [x] I have included tests that cover my changes.
- [x] All new and existing tests pass.

## Code review checklist
This is the checklist that the reviewer will follow while reviewing your pull request. You do not need to do anything with this checklist, but be aware of what the reviewer will be looking for.

- [ ] Ensure that any changes to external docs have been included in this pull request.
- [ ] If the changes require that minor/major versions need to be updated, tag the PR appropriately.
- [ ] Ensure the new code is [properly commented](https://github.com/golang/go/wiki/CodeReviewComments#doc-comments) and follows [Idiomatic Go](https://dmitri.shuralyov.com/idiomatic-go).
- [ ] Check that tests have been written and that they cover the new functionality.
- [ ] Run tests and ensure they pass.
- [ ] Build and run the code, performing any applicable integration testing.

### CODE REVIEW

**Comment by linhvo** [comment]
@travisturner: Michael is still working on delete broadcast handler (#639) and adding more tests, but I still create a PR for your review of what we've done so far. It has already covered #608, #610, #611, #638

File: index.go [hunk 1]
```diff
@@ -585,3 +611,113 @@ type importData struct {
...
+			return err
+		}
+		inputFiles, err := inputPath.Readdir(0)
+		for _, file := range inputFiles {
+			input, err := i.newInputDefinition(i.InputDefPath(), file.Name())
+			if err != nil {
+				return err
+			}
+			input.Open()
+			i.inputDefinitions[file.Name()] = input
```
Comment (travisturner) [inline index.go]: is there a way we can share some of the logic from `createInputDefinition` around frame creation such that on startup opening the inputDefinition will create any frames that don't yet exist?

File: index.go [hunk 2]
```diff
@@ -585,3 +611,113 @@ type importData struct {
...
+
+	inputDef.LoadDefinition(pb)
+	if err = inputDef.saveMeta(); err != nil {
+		return nil, err
+	}
+	i.inputDefinitions[pb.Name] = inputDef
+	return inputDef, nil
+}
+
+func (i *Index) newInputDefinition(path, name string) (*InputDefinition, error) {
```
Comment (travisturner) [inline index.go]: will there ever be a case where `path` is not `i.InputDefPath()`? if not, then we can remove the `path` argument from this function signature and just replace `i.name` with `i.InputDefPath()` inside the function.

**Comment by linhvo** [comment]
@travisturner: I removed WIP and it's ready for another review if necessary before sending another PR for /input handling

File: input_definition.go [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+}
+
+func (i *InputDefinition) AddFrame(frame InputFrame) error {
+	i.frames = append(i.frames, frame)
+	if err := i.saveMeta(); err != nil {
+		return err
+	}
+	return nil
+}
+
```
Comment (raskle) [inline input_definition.go]: Add comments to all public functions

File: handler.go [hunk 4]
```diff
@@ -1495,3 +1498,134 @@ func errorString(err error) string {
...
+				continue
+			} else {
+				err = fmt.Errorf("primary field's name not match columnLabel")
+				http.Error(w, err.Error(), http.StatusBadRequest)
+				return
+			}
+		}
+	}
+	if numPrimaryKey > 1 {
+		err = errors.New("duplicate primaryKey with other field")
```
Comment (travisturner) [inline handler.go]: let's make these errors part of the system error constants.
that way, when we check for them in tests, we don't have to compare the body of the error.

**Comment by travisturner** [comment]
@linhvo @raskle the test coverage on the `input-definition` branch is pretty good with the exception of error return cases. you might review those and try to test for the more likely cases.

File: handler.go [hunk 5]
```diff
@@ -1495,3 +1498,134 @@ func errorString(err error) string {
...
+	err := h.Broadcaster.SendSync(
+		&internal.DeleteInputDefinitionMessage{
+			Index: indexName,
+			Name:  inputDefName,
+		})
+	if err != nil {
+		h.logger().Printf("problem sending CreateInputDefinition message: %s", err)
+	}
+
+	if err := json.NewEncoder(w).Encode(postInputDefinitionResponse{}); err != nil {
```
Comment (linhvo) [inline handler.go]: since they are both empty struct, I will change it to `defaultInputDefinitionResponse` if it sounds ok

**Comment by travisturner** [comment]
@linhvo @raskle 
I refactored some of the `Encode()` logic to simplify it a bit, but we need to handle validation somewhere other than in `Encode()` (see #694). I don't think `Encode()` should return errors; we should check for errors prior to that. I think validation should happen either during or directly after the JSON decoding step.

**Pull request merged**